src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelThread.cs

```diff
@@ -98,67 +79,53 @@ public Task PostAsync(Action<object> callback, object state)
             return tcs.Task;
         }
 
-        private void PostCloseHandle(Action<IntPtr> callback, IntPtr handle)
+        private void ThreadStart(object parameter)
         {
-            lock (_workSync)
-            {
-                _closeHandleAdding.Enqueue(new CloseHandle { Callback = callback, Handle = handle });
-            }
-            _post.Send();
+            var tcs = (TaskCompletionSource<int>)parameter;
+            SetupLoop(tcs);
+            RunLoop();
         }
 
-        private void ThreadStart(object parameter)
+        private void SetupLoop(TaskCompletionSource<int> tcs)
         {
-            var tcs = (TaskCompletionSource<int>)parameter;
             try
             {
-                _loop.Init(_engine.Libuv);
-                _post.Init(_loop, OnPost);
+                _loop = new UvLoopHandle();
+                _post = new UvAsyncHandle(_loop, OnPost);
                 tcs.SetResult(0);
             }
             catch (Exception ex)
             {
                 tcs.SetException(ex);
             }
+        }
 
-            try
+        private void RunLoop()
+        {
+            using (_loop)
+            // _post is disposed in the uv_walk callback
             {
-                var ran1 = _loop.Run();
-                if (_stopImmediate)
+                try
                 {
-                    // thread-abort form of exit, resources will be leaked
-                    return;
+                    _post.Reference();
+                    _loop.Run();
+
+                    _loop.Validate();
+                    UnsafeNativeMethods.uv_walk(
```

Comment:
Hmm, you bring up a good point.

If I put a break point in the Dispose logic while there is an active SignalR connection, I still get `UvAsyncHandle.Send`-Requests that come in through `FrameResponseStream.Write`. I didn't look into that, but now that you brought it up I actually looked a the stack trace when this happens.

```
   at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at Microsoft.AspNet.Server.Kestrel.Networking.UvAsyncHandle.Send()
   at Microsoft.AspNet.Server.Kestrel.KestrelThread.Post(Action`1 callback, Object state)
   at Microsoft.AspNet.Server.Kestrel.Http.SocketOutput.Write(ArraySegment`1 buffer, Action`2 callback, Object state)
   at Microsoft.AspNet.Server.Kestrel.Http.Frame.Write(ArraySegment`1 data, Action`2 callback, Object state)
   at Microsoft.AspNet.Server.Kestrel.Http.FrameResponseStream.WriteAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at System.IO.Stream.WriteAsync(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.AspNet.Server.Kestrel.Http.FrameResponseStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.AspNet.SignalR.ResponseExtensions.Write(HttpResponse response, ArraySegment`1 data)
   at Microsoft.AspNet.SignalR.Transports.ServerSentEventsTransport.PerformKeepAlive(Object state)
   at Microsoft.AspNet.SignalR.Transports.ServerSentEventsTransport.<>c__DisplayClass0.<KeepAlive>b__1(Object state)
   at Microsoft.AspNet.SignalR.Infrastructure.TaskQueue.InvokeNext(Func`2 next, Object nextState, Object queueState)
   at Microsoft.AspNet.SignalR.Infrastructure.TaskQueue.<>c__DisplayClass0.<Enqueue>b__1(Func`2 n, Object ns, TaskQueue q)
   at Microsoft.AspNet.SignalR.TaskAsyncHelper.FromMethod[T1,T2,T3](Func`4 func, T1 arg1, T2 arg2, T3 arg3)
   at Microsoft.AspNet.SignalR.TaskAsyncHelper.Then[T1,T2,T3](Task task, Func`4 successor, T1 arg1, T2 arg2, T3 arg3)
   at Microsoft.AspNet.SignalR.Infrastructure.TaskQueue.Enqueue(Func`2 taskFunc, Object state)
   at Microsoft.AspNet.SignalR.Transports.TransportDisconnectBase.EnqueueOperation(Func`2 writeAsync, Object state)
   at Microsoft.AspNet.SignalR.Transports.ServerSentEventsTransport.KeepAlive()
   at Microsoft.AspNet.SignalR.Transports.TransportHeartbeat.CheckTimeoutAndKeepAlive(ConnectionMetadata metadata)
   at Microsoft.AspNet.SignalR.Transports.TransportHeartbeat.Beat(Object state)
[...]
   at System.Threading.TimerQueue.AppDomainTimerCallback()"
```

Looks like the heartbeat does not stop.